### PR TITLE
add curried MuxLookup.apply, deprecate old apply

### PR DESCRIPTION
**Pull request by albertchen-sifive**
fixes https://github.com/chipsalliance/chisel/issues/2780

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

new feature/API

#### API Impact

deprecates the current `MuxLookup.apply` and adds two overloads which behave the same as the old `MuxLookup.apply` but take two parameter lists instead. `MuxLookup.fromEnum` is replaced with the new `MuxLookup.apply`.

#### Backend Code Generation Impact

none, but should generate better source locators for `MuxLookup.apply`

#### Desired Merge Strategy

Squash

#### Release Notes

Add a new version of `MuxLookup.apply` that takes two parameter lists instead of one. This helps the scala compiler report better type errors.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x` or `3.6.x` depending on impact, API modification or big change: `5.0.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/main/scala/chisel3/util/Mux.scala [hunk 1]
```diff
@@ -49,21 +52,47 @@ object PriorityMux {
...
+  def apply[T <: Data](key: UInt, default: T)(mapping: Seq[(UInt, T)]): T =
+    macro MuxLookupTransform.applyCurried[UInt, T]
+
   /** @param key a key to search for
     * @param default a default value if nothing is found
     * @param mapping a sequence to search of keys and values
     * @return the value found or the default if not
     */
-  def apply[S <: UInt, T <: Data](key: S, default: T, mapping: Seq[(S, T)]): T = {
+  def apply[T <: Data](key: EnumType, default: T)(mapping: Seq[(EnumType, T)]): T =
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/Mux.scala:92]: Shouldn't this still have some type E that extends an enumtype? Like I should not be able to mix days of the week with months

File: src/test/scala/chiselTests/MuxSpec.scala [hunk 2]
```diff
@@ -43,14 +43,14 @@ class MuxLookupEnumTester extends BasicTester {
...
+  assert(MuxLookup(TestEnum.b, 3.U)(mapping) === 1.U)
+  assert(MuxLookup(TestEnum.c, 3.U)(mapping) === 2.U)
 
   val incompleteMapping = Seq(TestEnum.a -> 0.U, TestEnum.c -> 2.U)
-  assert(MuxLookup.fromEnum(TestEnum.a, 3.U)(incompleteMapping) === 0.U)
-  assert(MuxLookup.fromEnum(TestEnum.b, 3.U)(incompleteMapping) === 3.U)
-  assert(MuxLookup.fromEnum(TestEnum.c, 3.U)(incompleteMapping) === 2.U)
+  assert(MuxLookup(TestEnum.a, 3.U)(incompleteMapping) === 0.U)
+  assert(MuxLookup(TestEnum.b, 3.U)(incompleteMapping) === 3.U)
+  assert(MuxLookup(TestEnum.c, 3.U)(incompleteMapping) === 2.U)
```
Comment (mwachs5) [inline src/test/scala/chiselTests/MuxSpec.scala:53]: Like a case where I define anotherEnum I should not be able to put AnotherEnum.c in for the default or any of the cases, I think.

**Review by mwachs5 — COMMENTED** [review]
Are you intentionally relaxing the constraint that the enumerated should all be of the same type?

File: src/main/scala/chisel3/util/Mux.scala (hunk 1 above)
Comment (albertchen-sifive) [reply src/main/scala/chisel3/util/Mux.scala:92]: ah yeah that's still a useful constraint to have

File: src/test/scala/chiselTests/MuxSpec.scala (hunk 2 above)
Comment (jackkoenig) [reply src/test/scala/chiselTests/MuxSpec.scala:48]: Agreed--we really need a testing flow for negative compile tests...

I have a prototype of the build flow to make this work, but we need some pithy way to express what we expect. Either a standalone `FileCheck`, a simpler custom similar thing, or perhaps Linux `expect`

File: src/main/scala/chisel3/util/Mux.scala [hunk 3]
```diff
@@ -49,21 +52,47 @@ object PriorityMux {
...
+    macro MuxLookupTransform.applyCurried[UInt, T]
+
   /** @param key a key to search for
     * @param default a default value if nothing is found
     * @param mapping a sequence to search of keys and values
     * @return the value found or the default if not
     */
-  def apply[S <: UInt, T <: Data](key: S, default: T, mapping: Seq[(S, T)]): T = {
+  def apply[S <: EnumType, T <: Data](key: S, default: T)(mapping: Seq[(S, T)]): T =
+    apply[T](key.asUInt, default)(mapping.map { case (s, t) => (s.asUInt, t) })
```
Comment (jackkoenig) [inline src/main/scala/chisel3/util/Mux.scala:93]: Can we have a source locator on this one too?

File: src/test/scala/chiselTests/MuxSpec.scala (hunk 2 above)
Comment (albertchen-sifive) [reply src/test/scala/chiselTests/MuxSpec.scala:48]: we could use a macro like the one shapeless uses https://github.com/milessabin/shapeless/blob/main/core/shared/src/main/scala/shapeless/test/typechecking.scala

Comment (jackkoenig) [reply src/test/scala/chiselTests/MuxSpec.scala:48]: ScalaTest has something similar, but macros don't work super well because they aren't full separate invocations of the compiler and thus can't take into account differing compile options

**Comment by jackkoenig** [comment]
Since this deletes an API add in https://github.com/chipsalliance/chisel/pull/3071 and that was backported to 3.6.x and 3.5.x, we need to backport this. As is, this won't be binary compatible because of the change to the deprecated apply method, but that can be fixed in the 3.5.x backport (we can just remove the use of the macro and it won't have a good source locator--it's fine).

**Review by jackkoenig — APPROVED** [review]
LGTM! Nice!

**Comment by jackkoenig** [comment]
I pushed a commit that makes sure the `do_` methods are properly segregated in the ScalaDoc to make it cleaner.
